Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm

```diff
@@ -1625,27 +1634,76 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
-    menuItems.append({ WebCore::contextMenuItemPDFSinglePageContinuous(), currentDisplayMode == ContextMenuItemTag::SinglePageContinuous, enumToUnderlyingType(ContextMenuItemTag::SinglePageContinuous), ContextMenuItemEnablement::Enabled, ContextMenuItemHasAction::Yes, ContextMenuItemIsSeparator::No });
-    menuItems.append({ WebCore::contextMenuItemPDFTwoPages(), currentDisplayMode == ContextMenuItemTag::TwoPages, enumToUnderlyingType(ContextMenuItemTag::TwoPages), ContextMenuItemEnablement::Enabled, ContextMenuItemHasAction::Yes, ContextMenuItemIsSeparator::No });
-    menuItems.append({ WebCore::contextMenuItemPDFTwoPagesContinuous(), currentDisplayMode == ContextMenuItemTag::TwoPagesContinuous, enumToUnderlyingType(ContextMenuItemTag::TwoPagesContinuous), ContextMenuItemEnablement::Enabled, ContextMenuItemHasAction::Yes, ContextMenuItemIsSeparator::No });
+    menuItems.appendVector(displayModeContextMenuItems());
 
     addSeparator();
 
-    menuItems.append({ WebCore::contextMenuItemPDFZoomIn(), 0, enumToUnderlyingType(ContextMenuItemTag::ZoomIn), ContextMenuItemEnablement::Enabled, ContextMenuItemHasAction::Yes, ContextMenuItemIsSeparator::No });
-    menuItems.append({ WebCore::contextMenuItemPDFZoomOut(), 0, enumToUnderlyingType(ContextMenuItemTag::ZoomOut), ContextMenuItemEnablement::Enabled, ContextMenuItemHasAction::Yes, ContextMenuItemIsSeparator::No });
-    menuItems.append({ WebCore::contextMenuItemPDFActualSize(), 0, enumToUnderlyingType(ContextMenuItemTag::ActualSize), ContextMenuItemEnablement::Enabled, ContextMenuItemHasAction::Yes, ContextMenuItemIsSeparator::No });
+    menuItems.appendVector(scaleContextMenuItems());
 
-    return { contextMenuPoint, WTFMove(menuItems), { enumToUnderlyingType(ContextMenuItemTag::OpenWithPreview) } };
+    auto contextMenuPoint = frameView->contentsToScreen(IntRect(frameView->windowToContents(contextMenuEvent.position()), IntSize())).location();
+
+    return PDFContextMenu { contextMenuPoint, WTFMove(menuItems), { enumToUnderlyingType(ContextMenuItemTag::OpenWithPreview) } };
+}
+
+PDFContextMenuItem UnifiedPDFPlugin::separatorContextMenuItem() const
+{
+    return { { }, 0, enumToUnderlyingType(ContextMenuItemTag::Invalid), ContextMenuItemEnablement::Disabled, ContextMenuItemHasAction::No, ContextMenuItemIsSeparator::Yes };
+}
+
+Vector<PDFContextMenuItem> UnifiedPDFPlugin::selectionContextMenuItems(const IntPoint& contextMenuPointInPluginSpace) const
+{
+    if (![m_pdfDocument allowsCopying] || !m_currentSelection)
+        return { };
+
+    Vector<PDFContextMenuItem> items {
+        { contextMenuItemTagSearchWeb(), 0, enumToUnderlyingType(ContextMenuItemTag::WebSearch), ContextMenuItemEnablement::Enabled, ContextMenuItemHasAction::Yes, ContextMenuItemIsSeparator::No },
```

Comment:
These are pretty unwieldy and all have the same last 3 arguments. Consider making a helper function to create them.

Reply:
Yes good point. Let me see

Reply:
Simplified a bunch in [4a4ff6d](https://github.com/WebKit/WebKit/commit/4a4ff6d5ebd89f3049e57f31634b35865defc8a4)

Reply:
https://github.com/WebKit/WebKit/pull/24392/files#diff-ad530d257d17aa633a189e9b9afb27cdb009a5d81cbf7f42c927d8a176d8464cR1683-R1695

```cpp
PDFContextMenuItem UnifiedPDFPlugin::contextMenuItem(ContextMenuItemTag tag) const
{
    switch (tag) {
    case ContextMenuItemTag::Unknown:
    case ContextMenuItemTag::Invalid:
        return separatorContextMenuItem();
    default: {
        auto currentDisplayMode = contextMenuItemTagFromDisplayMode(m_documentLayout.displayMode());
        int state = isDisplayModeContextMenuItemTag(tag) ? currentDisplayMode == tag : 0;
        return { titleForContextMenuItemTag(tag), state, enumToUnderlyingType(tag), ContextMenuItemEnablement::Enabled, ContextMenuItemHasAction::Yes, ContextMenuItemIsSeparator::No };
    }
    }
}
```